symengine/ntheory.cpp

```diff
@@ -1474,4 +1474,69 @@ void powermod_list(std::vector<RCP<const Integer>> &pows, const RCP<const Intege
     }
 }
 
+std::vector<mpz_class> quadratic_residues(const Integer &a)
+{
+    /*
+        Returns the list of quadratic residues.
+        Example
+        ========
+        >>> quadratic_residues(7)
+        [0, 1, 2, 4]
+    */
+
+    if (a.as_mpz() < 1) {
+        throw std::runtime_error("quadratic_residues: Input must be > 0");
+    }
+
+    std::vector<mpz_class> residue;
+    for(mpz_class i = 0; i <= a.as_int()/2; i++)
+    {
+        residue.push_back((i*i) % a.as_int());
+    }
+
+    sort(residue.begin(), residue.end());
+    residue.erase(unique(residue.begin(), residue.end()), residue.end());
+
+    return residue;
+
+}
+
+bool is_quad_residue(const Integer &a, const Integer &p)
+{
+    /*
+    Returns true if ``a`` (mod ``p``) is in the set of squares mod ``p``,
+    i.e a % p in set([i**2 % p for i in range(p)]). If ``p`` is an odd but
+    not prime, an iterative method is used to make the determination.
+    */
+
+    mpz_class p2 = p.as_mpz();
+    if (p2 < 1)
+        throw std::runtime_error("is_quad_residue: Second parameter must be > 0");
+
+    mpz_class a_final = a.as_mpz();
+    if (a.as_mpz() >= p2 || a.as_mpz() < 0)
+        a_final = a.as_mpz() % p2;
+    if (a_final < 0 && a_final > -p2)
+        a_final += p2;
+    if (a_final < 2 || p2 < 3)
+        return true;
+
+    if (!probab_prime_p(*integer(p2), 0)){
```

Comment:
Do you mean
```int r;
if (!probab_prime_p(*integer(p2), r)){
```
or should I use something like
`if (!probab_prime_p(*integer(p2), 50)){`
since that would be more accurate